Add blog post about Planetscape (#5)
social-dist0rtion-protocol/www-data: source/_posts/Planetscape-a-dystopian-escape-game-for-36C3.md

---
title: Planetscape
date: 2020/01/30
---

# Planetscape, a dystopian escape game for 36C3

![][img:header]

<!--

> ## Summary

Reviewer: Bit confused about this. Is hexo displaying this somehow?

Author: It's rendered in the HTML page but as a comment. It was the summary but I'm not sure it's needed. Would you keep it as summary or not?

- A web dApp that players use to read the story, submit passwords, and check the leaderboard.
- A smart contract to store the structure of the story, validate players' submissions, and read the leaderboard.
- Encrypted JSON documents stored in IPFS that contain the actual text of the chapters.

We pledged to create a *user-friendly application that would rely on decentralized technology only*. We want to live the decentralized dream with no compromises, developing a powerful dApp that would use Ethereum under the hood without exposing any details about blockchain and IPFS to the player. We committed to have **instant onboarding** by allowing people to play the game just by visiting a website.

We achieved that by leveraging our learnings from [Hunt Zuckerberg][3] and [Planet A][4] by using a "burner-wallet"-like approach (the player's wallet is created on the fly when opening the dApp for the first time). In order to make transactions players needed Ether. Given that the game was played on the [Rinkeby][6] network, we set up a *gas station* that would simply [transfer][7] a small amount of Ether to new players.

Leveraging blockchain technology and IPFS give us many advantages:

> - It allows us not to manage any server.

Reviewer: ```suggestion
- We don't have to manage servers
```

> - Each time a correct solution is found, a transaction is sent to the smart contract of the game. This feature allows us to easily create stats on the number of players and the difficulty of each level.

Reviewer: ```suggestion
- Each time a player finds a correct solution, a transaction is sent to the smart contract of the game. This allows us to easily create stats on the number of players and the difficulty of each level.
```

- The smart contract of the game maps each chapter to its IPFS content hash, once the game is set up there is no way for the game masters to cheat.
- (For the future) include other game masters to expand the main story or create sub-stories.

### Introducing THC
In order to orchestrate all different components of the game, Social Dist0rtion Protocol (yes, still us) developed a game engine called [Treasure Hunt Creator][0] (**THC** for short) that, given a story structured in directories, is able to:

- Build and deploy the front-end application.
- Build, deploy and initialize the game smart contract on the Ethereum Network.
- Encrypt and deploy the chapters on IPFS.

[THC][0] played a crucial role in the creation of the game, allowing us to meet our high standards in terms of development experience. But THC was more than that. Fast prototyping and automatic deploying gave us the piece of mind we needed during our last hours of work.

### Zero Knowledge Proof

> As mentioned earlier, every time a correct password is found, a transaction is sent to the smart contract managing the game. Ethereum transactions are public, and we don't want players to cheat sniffing passwords submitted by others. To avoid that, we implemented a cheap zero-knowledge proof scheme. Roughly speaking, the scheme is the following:

Reviewer 2: ```suggestion
As mentioned earlier, every time a correct password is found, a transaction is sent to the smart contract. Ethereum transactions are public and we don't want players to sniff passwords submitted by others. To avoid that, we implemented a cheap zero-knowledge proof scheme. Roughly speaking, the scheme is the following:
```

1. Given a valid `password`, a player's `player_address`:
2. The dApp generates `seed = keccak256(password)`.
3. The dApp generates `chapter_private_key` from `seed`.
4. The dApp generates `signature = sign(address, chapter_private_key)`

> 5. The dApp calls the `submit` method of the Smart Contract passing `signature`.

Reviewer: ```suggestion
5. The dApp calls the `submit` method of the smart sontract passing `signature`.
```

> 6. The Smart Contract checks the current `chapter` for `player_address` in its storage.

Reviewer: ```suggestion
6. The smart contract checks the current `chapter` for `player_address` in its storage.
```

> 7. The Smart Contract loads the `chapter_address` for `chapter`.

Reviewer: ```suggestion
7. The smart contract loads the `chapter_address` for `chapter`.
```

> 8. The Smart Contracts calculates `signing_address` by doing an `ecrecover(player_address, signature)`.

Reviewer: ```suggestion
8. The smart contracts calculates `signing_address` by doing an `ecrecover(player_address, signature)`.
```

> 9. If `signing_address == chapter_address`, then the Smart Contract updates the current chapter for the player, moving them to the next chapter.

Reviewer: ```suggestion
9. If `signing_address == chapter_address`, then the smart contract updates the current chapter for the player, moving them to the next chapter.
```

More information about the [creation][10] and [verification][11] of the proof can be found in the source code.

## Marketing

> Given the importance of **operation Planetscape** we wanted to onboard as many people as possible. We are not acquainted with custom and traditions of the early 2000<sup id="a1">[1](#f1)</sup> (remember, we are from the future) so we tried three orthogonal strategies to engage with attendees of the Congress.

Reviewer 2: ```suggestion
Given the importance of **operation Planetscape**, we wanted to onboard as many people as possible. We are not acquainted with custom and traditions of the early 2000<sup id="a1">[1](#f1)</sup> (remember, we are from the future) so we tried three orthogonal strategies to engage with attendees of the Congress.
```

### Personal Envelope
Before reaching the Congress we handcrafted 35 personal letters and embedded them in envelopes<sup id="a2">[2](#f2)</sup>. Each letter contained the text with the call to action to open the website, start the game, and enter the password for *Chapter 0*. Three SDP members carried them around the congress to personally onboard attendees of 36C3. Handing the envelope was **not** a trivial task and required interaction to assess the willingness of the prospect to participate to the game. Remember, we had a limited number of envelopes, so we had to play them carefully.

> 

Reviewer 2: do you have a picture of the letter to embed here?

Author: Not of the envelop :( but I have a couple of more pictures form the congress to share.

## Results

Like we did for Hunt Zuckerberg, before starting the game we had some goals in mind in terms of:

- number of users that plays the game
- number of users that finishes the game

At the time of the first treasure hunt, we wanted to have at least 5 players (not including us) and at least 1 person finishing the game. We achieved the astounding result of 10 players (including 3 of us) and 3 players running for the final piece of the puzzle! Two of them are now honoured members of SDP (we love you Dina and Donald).

With Planetscape we then decided to set the bar higher: at least 100 players and 2 people finishing the game.

Our optimism was suggesting us to not expect anything more than 20 players, but given the aggressive marketing campaign and the complete lack of setup costs in order to start playing the game, we though 100 would have actually been a reasonable number.

Final results:

- 254 player (excluding us) joined the game and finished **at least Chapter 0**. You can check the [list at the leaderboard](https://adventure.dist0rtion.com/#/leaderboard).
- 3 players were running in parallel for the first price, it was really exciting to see that.

> - Ultimately, 12 player finished the game.

Reviewer: ```suggestion
- Ultimately, 12 players finished the game.
```

![][img:difficulty]

It seems like the most annoying chapter, for the players that got there, was the one involving the hunt of an access point at the congress. In fact, at some point, someone suddenly disconnected it and we found it out only an hour later. The general feedback was that that chapter was a pain in the ass.

## What's next

On a technical side, Social Dist0rtion Protocol will continue the development of THC:

- Enable easier customization and theming of the dApp.
- Reduce technical debt introduced by last minute changes.
- Improve the overall documentation.
- Improve the IPFS part as well: right now the dApp fetches the assets from the centralized infura IPFS gateway. By using [js-ipfs][8] we can overcome this limitation and have a purely decentralized approach. There is still no known solution to avoid using centralized JSONRPC like infura.

Given the feedback received, Social Dist0rtion Protocol is committed to repeat this operation for other events by creating more engaging stories and puzzles.

## Stay up to date!

We have a nice newsletter and you should subscribe to it.

<hr/>

*Footnotes from the Future*

> <small><a name="f1">1</a>: Custom and traditions of humans of the early 2000 are not very well known in the year 2111. Most of the evidence and documents have been destroyed in *The Great DataCenter Meltdown™* dated 2051. Billions and billions and billions of relevant documents like instagram selfies, tiktok videos, medium posts, memes, and tweets had been lost forever in this tragic event. That year is also remembered as *The Influcene Extinction*, where millions of influencers had to quit social media and find a real job.[↑](#a1)</small>

Reviewer: ```suggestion
<small><a name="f1">1</a>: Custom and traditions of humans of the early 2000 are not very well known in the year 2111. Most of the evidence and documents have been destroyed in *The Great DataCenter Meltdown™* dated 2051. Billions and billions and billions of relevant documents like Instagram selfies, Tiktok videos, Medium posts, memes, and tweets had been lost forever in this tragic event. That year is also remembered as *The Influence Extinction*, where millions of influencers had to quit social media and find a real job.[↑](#a1)</small>
```
